### PR TITLE
Add CMake option to enable address sanitizer checks

### PR DESCRIPTION
**Pull request by phoerious**
## Description

This patch adds a CMake option `-DWITH_ASAN` to enable GCC/Clang address sanitizer checks. It also lists an address sanitizer check as a required merge prerequisite in the PR template checklist.

## Motivation and Context

Considering that pull requests keep adding third-party API dependencies which are written in C, it becomes more likely that memory management can go wrong and buffer overflows or double frees etc. occur. Running an address sanitizer as a required check before submitting a PR (or merging it) is therefore a good idea.

## How Has This Been Tested?

I compiled the code with GCC on Ubuntu 16.10. I haven't yet tested it with Clang on my Arch system at home. Depending on the compiler or platform, it may be necessary to disable ODR checks, which may or may not be sensible to do directly in the code (instead of using the ASAN_OPTIONS env variable).

## Types of changes

- ✅ New feature (non-breaking change which adds functionality)

## Checklist:

- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**

### CODE REVIEW

**Comment by phoerious** [comment]
After a little more tinkering and testing, I also managed to enable ASAN for our Travis build, but had to disable memory leak detection, because lots of third-party libs are generating leak errors which cause our tests to fail. So to detect leaks, a manual run is needed. For the main program (not the tests), I reduced the memory leak error dump at the end as much as possible by forcing the leak check to run before the program quits. On my Ubuntu 16.10 system here, one error in libfontconfig.so is still there (probably a real memory leak, not only one where the developer thought that using free() at the end of the program is not necessary), but that's better than 100 or more.

For Clang, I need to reduce the ODR violation check level from 2 to 1, otherwise our Auto-Type lib throws an error.

**Comment by phoerious** [comment]
I tested it on Windows and OS X now. Windows doesn't support -fsanitize=address at all, OS X only does not support -fsanitize=leak, but -fsanitize=address somehow also doesn't do what I expect it to do and generates a lot of warnings and other weird messages during compilation. I therefore added an error message to CMake when trying to use -DWITH_ASAN=ON on a platform other than Linux.

File: README.md [hunk 1]
```diff
@@ -58,14 +58,15 @@ sudo make install
...
+  -DWITH_XC_AUTOTYPE=[ON|OFF] Enable/Disable Auto-Type (default: ON)
+  -DWITH_XC_HTTP=[ON|OFF] Enable/Disable KeePassHTTP and custom icon downloads (default: OFF)
+  -DWITH_XC_YUBIKEY=[ON|OFF] Enable/Disable YubiKey HMAC-SHA1 authentication support (default: OFF)
 
   -DWITH_TESTS=[ON|OFF] Enable/Disable building of unit tests (default: ON)
   -DWITH_GUI_TESTS=[ON|OFF] Enable/Disable building of GUI tests (default: OFF)
-  -DWITH_DEV_BUILD=[ON|OFF] Enable/Disable deprecated method warnings. (default: OFF)
-  -DWITH_COVERAGE=[ON|OFF] Enable/Disable coverage tests. (GCC ONLY) (default: OFF)
+  -DWITH_DEV_BUILD=[ON|OFF] Enable/Disable deprecated method warnings (default: OFF)
+  -DWITH_ASAN=[ON|OFF] Enable/Disable address sanitizer checks (default: OFF)
```
Comment (TheZ3ro) [inline README.md]: Maybe here we can add a `(Linux ONLY)` like for `WITH_COVERAGE GCC ONLY`

Comment (phoerious) [inline README.md]: Done.

**Pull request merged**